### ISSUE TITLE
dotnet restore has troubles with proxy after 2.1 upgrade

### ISSUE THREAD

**Issue by rrelyea** [dotnet/corefx#30354]
Anything NuGet should have to do differently in 2.1 compared to 2.0?

The last several days, we've had several customers chime in on: https://github.com/NuGet/Home/issues/5358

For example:
@kukjevov said:
Hmmm interesting is that after setting http_proxy directly into c:\Users\<user>\AppData\Roaming\NuGet\NuGet.Config it started working, but setting http_proxy to local Nuget.Config which is in project it was not working.
Problem occured only after installing .net core SDK 2.1, before this i was able to work without http_proxy

@delebru said:
I'm getting the same issues unless I run dotnet restore as root on my linux server. I recently installed the 2.1 SDK and currently have SDKs 2.0 and 2.1 installed.
I did not use to have this problem when only SDK 2.0 was installed. I'm wondering if there's some permissions issue now with the new SDK installation?

and then later said:
sudo chmod +r /etc/pki/tls/certs/* solved the problem for me.

/cc @mishra14

**Comment by davidsh** [comment]
This is most likely a dupe of #30327 or #30330

**Comment by karelz** [comment]
@davidsh I am puzzled how certs permissions (case from @delebru) would be caused by our gaps in proxy support. Do you have a theory?

The case from @kukjevov looks like HTTP proxy is not automatically detected and has to be explicitly set by nuget via config. That could be explained by one of the linked bugs.

**Comment by davidsh** [comment]
>@davidsh I am puzzled how certs permissions (case from @delebru) would be caused by our gaps in proxy support. Do you have a theory?

It's possible that there are different root causes.  https://github.com/NuGet/Home/issues/5358 is actually an older issue that describes a general problem with NuGet restore and has been opened/closed a bunch of time.

We really need to see a crisp repro of this with a detailed description of the network environment with information like:

* Windows or Linux
* VM or Docker container
* proxy or no proxy
* authentication for proxy and schemes used (Negotiate, NTLM, etc.)
* method by which any username/password is passed to proxy (or via "default credentials")
* Wireshark traces would be useful too.

**Comment by delebru** [comment]
In my case the issue was on a dedicated linux server with no docker nor proxy. Only the root user was able to restore nuget packages.

**Comment by karelz** [comment]
I was about to ask @rrelyea to untangle the original NuGet issue - I didn't notice it is that old. 2017 reports and .NET Core 2.0 reports are unrelated to this issue. Only .NET Core 2.1 reports may be related to the known proxy issues linked above.
@rrelyea I think we need to treat each report separately and first find root-cause of various symptoms. How do you want to do that on the original NuGet issue?

@delebru's report seems to be about permissions - I guess 2.1 changed how it accesses certs by using OpenSSL directly, not via libcurl. That, however, does not explain permissions IMO - even if you use libcurl, it is still the same process with same permissions limitations.
@delebru are you able to revert back into the "broken state" and do some experiments? (to help us root-cause the problem)

@kukjevov are you able to run Wireshark on your failing repro? That way we can check if your situation is dupe of one of the issues.

**Comment by delebru** [comment]
Unfortunately I can't remember all permissions in that folder, but I'm certainly keen to do some tests if helps.

**Comment by karelz** [comment]
@delebru were you able to reproduce it on another machine/VM/docker image? If there's no repro we can dig into, there is not much we can do I'm afraid.

**Comment by kukjevov** [comment]
Hi i havent tried Wireshark since it started working for me. But i can if it helps. So should i, let me know.

**Comment by karelz** [comment]
We need traces from the failed case. If you can collect those, that would be useful.

**Comment by tebeco** [comment]
added in my agenda for tomorrow morning at work

**Comment by tebeco** [comment]
@nkolev92 is there any description on how/what you want from wireshark ?
also do you want trace from 2.1.300 or 301 ? or both ?

**Comment by nkolev92** [comment]
@tebeco, I think @karelz should be able to help with the details. 

I think either 2.1.300 or 301 works, since I presume both would have the same issue.

**Comment by karelz** [comment]
@wfurt can you please look at it? This seems to impact quite a few people
cc @stephentoub @geoffkizer

**Comment by tebeco** [comment]
that would be a really good idea to add end to end test for proxy scenario that make it impossible to push an RTM breaking proxy basic behavior

or it should be an OPT IN, not optOut
especially if you want adoption for companies

i’m doing politics with interns so that our company pays to have nuget proxy internal to reach nuget.org
this is a bit crazy right ?

**Comment by OnamChilwan** [comment]
Is there a temporary workaround anybody is aware of? Even reverting to an older version..

**Comment by wfurt** [comment]
I agree with @davidsh , this seems like two different issues. The Linux branch seems to have troubles with CA certificates for  SSL. What exact  OS do you use @delebru? I can try to reproduce it. 
Conceptually since old HTTP openssl/curl runs in same process the steps should fail same way. 

The windows part seems related specifically to NTLM proxy auth. Setting DOTNET_SYSTEM_NET_HTTP_USESOCKETSHTTPHANDLER=0 should fix it. 
If anybody has indication that it also fails with Basic or Digest, please provide some evidence.

**Comment by delebru** [comment]
It happened to me on a Centos 7 but I tried to reproduce it yesterday on a fresh install and it seemed to work out of the box. 

On the system that failed, I also have running a Plesk web panel. I was running some apps with the 2.0 sdk with no issues for a while. I've only had those failures after installing version 2.1.300.

**Comment by wfurt** [comment]
On my Centos7 I have  /etc/ssl/certs/ and that symlinks to /etc/pki/ca-trust/extracted/pem - all readable by everybody. Now, Centos is interesting case as curl by default links with nss instead of OpenSssl. That can possibly use different cert storage. To see that you can write simple app to simply fetch HTTPS URL using HttpClient.  That would show us is the problem is with proxy or underlying SslStream.

BTW Using HttpClient may be good test even on the windows cases.

**Comment by Flern** [comment]
@wfurt 
I confirmed that `set DOTNET_SYSTEM_NET_HTTP_USESOCKETSHTTPHANDLER=0` resolved the proxy issue for me on Windows Server 2016 running Docker containers for builds.

Thank you!

**Comment by wfurt** [comment]
Do you know if your proxy uses NTLM @Flern ?

**Comment by Flern** [comment]
I'm not 100% positive - it's a corporate proxy.  Any easy way I can check to see if it's using NTLM @wfurt ?

**Comment by Flern** [comment]
My 407 response on failure indicates:
```
HTTP/1.1 407 Proxy Authentication Required
Proxy-Authenticate: NEGOTIATE
Proxy-Authenticate: NTLM
```
Successful runs also show the NTLM activity (NTLM Secure Service Provider with negotiate flags, etc.).

**Comment by davidsh** [comment]
@Flern What are the rest of the response headers?  If it contains "Connection: close", then this is issue #30327.   If not, then if you're final destination is HTTPS server (thru the proxy), then this is issue #30330.

**Comment by Flern** [comment]
@davidsh Yes, the 407 response includes "Connection: close"

**Comment by davidsh** [comment]
>@davidsh Yes, the 407 response includes "Connection: close"

Then the problem is issue #30327

**Comment by karelz** [comment]
Duplicate of #30327

If there are other root-causes mixed up in this report, please let us know. To tell for sure, Wireshark traces need to be verified. Thanks!
